content/python/concepts/decorators/decorators.md

```diff
@@ -0,0 +1,482 @@
+
+```py
+def general_decorator(original_function):
+
+    # accepting variable number of arguments
+    def wrapper(*args, **kwargs):
+        func(args, kwargs)
+
+    return wrapper
+
+@general_decorator
+def original_function(args, kwargs):
+    # code for the function
+```
+
+An example of variable arguments accepting decorator is as follows:
+
+```codebyte/py
+def upper_case_decorator(original_function):
+
+    def wrapper(*arg1, **kwargs):
+
+        result = original_function(*arg1)
+
+        result = result.upper()
+        return result
+    return wrapper
+
+
+@upper_case_decorator
+def greeting(arg1):
+    return arg1 + "!"
+
+
+@upper_case_decorator
+def double_greeting(arg1, arg2):
+    return arg1 + ". " + arg2 + "!"
+
+
+print(greeting('hi'))
+# output:
+# HI!
+
+print(double_greeting('hi', 'hello'))
+# output:
+# HI. HELLO!
+
+```
+
+## Passing Arguments to Decorators
+
+Arguments can also be passed to the decorator itself. To pass arguments to the Decorator, there should be a `decorator_wrapper_to_accept_decorator_arguments()` function wrapping the original decorator.
+
+```py
+def decorator_wrapper_to_accept_decorator_arguments(decorator_arg1, decorator_arg2):
+
+    def decorator(original_function):
+
+        def wrapper(*args, **kwargs):
+            original_function(args, kwargs)
+
+        return wrapper
+
+    # (IMPORTANT)
+    return decorator
+
+@decorator_wrapper_to_accept_decorator_arguments('value 1', 'value 2')
+def function_name(arg):
+  # statements
+
+```
+
+Eg: With the greeting function, the decorator could be setup to use a `number_of_times_to_greet` parameter to greet once or more than once.
+
+```py
+def upper_case_decorator_wrapper(number_of_times_to_greet):
+    def upper_case_decorator(original_function):
+
+        def wrapper(*arg1, **kwargs):
+
+            result = ""
+            for i in range(number_of_times_to_greet):
+                result += original_function(*arg1, **kwargs)
+            result = result.upper()
+            return result
+        return wrapper
+
+    return upper_case_decorator
+
+
+@upper_case_decorator_wrapper(1)
+def greeting(arg1):
+    return arg1 + "!"
+
+
+@upper_case_decorator_wrapper(2)
+def greeting_twice(arg1):
+    return arg1 + "!"
+
+
+print(greeting('hi'))
+# output:
+# HI!
+
+print(greeting_twice('hi'))
+# output:
+# HI!HI!
+
+```
+
+Even though, `greeting()` and `greeting_twice()` have the exact same function definition, their outputs are different because the decorator is executing these functions as per the `number_of_times_to_greet` argument passed to the decorator.
+
+## `help()` with Decorators
+
+The original function name and docstring are lost when it's wrapped by a decorator.
+
+```codebyte/py
+def upper_case_decorator(original_function):
+
+    def wrapper():
+        """decorates a function by uppercasing the return value"""
```

Comment:
Could we change this to a single line comment? I believe the `"""` syntax is used for multi line comments.
```suggestion
        # decorates a function by uppercasing the return value
```

Reply:
This is not a single line comment. That's not what the purpose of `"""` is. They represent docstrings in python and not multi-line comments. Using the triple double quotes, programmers can write docstrings in python to associate documentation with the code. Docstrings can also be single-line or multi-line. I'm resolving this change and cascading it to other same suggestions as well.

Reply:
@Goku-kun Fair point. I forgot about that. It was late

Reply:
No worries. We are a community. We help each other. With stuff like this, everyone comes out ahead. xD